### PR TITLE
Allow points to be awarded for participation

### PR DESCRIPTION
**Pull request by muyjohno**
Points are awarded as defined in league settings for each game played
Participation points can be capped in total

Also refactored some rule and ruleset methods
Added participation and achievement points to leaderboard

### CODE REVIEW

File: app/models/rule.rb [hunk 1]
```diff
@@ -8,11 +8,13 @@ class Rule < ActiveRecord::Base
     points_for_win:       0,
     points_for_loss:      1,
     points_for_tie:       2,
-    points_for_time_win:  3
```
Comment (richdouglasevans) [inline app/models/rule.rb]: Have you considered using a reduce function here?

Every time a new value is added the programmer has to add both a key (name) _and_ an index that is incremented by `n`. Adding the index is an extra unnecessary step.

The index is an implementation detail that has no bearing on the domain: I can't go to James and say "points for a win has a 0 index!" 'cos the index has nothing to do with Netrunner; I _can_ say "we've got points for a win, a tie, a loss and stuff like that" 'cos it's the names are the important thing and they do have meaning in the domain of Netrunner.

```ruby
KEYS = [
  :points_for_win,
  :points_for_loss,
  :points_for_tie,
  :points_for_time_win,
  :points_for_participation,
  :max_points_for_participation
].each_with_index.inject({}) { |memo, (key, index) | memo[key] = index; memo }
```

Using a reduction here means that when a new value is required the programmer just needs to add a new name to the source list: the index will be automatically derived.

File: app/models/rule.rb [hunk 2]
```diff
@@ -8,11 +8,13 @@ class Rule < ActiveRecord::Base
...
+    points_for_time_win:  3,
```
Comment (richdouglasevans) [inline app/models/rule.rb]: Tools that hound me about stuff that I don't care about makes me :crying_cat_face: 

Burn that setting in your Hound configuration. Do it for me Johno :smiley:

**Comment by muyjohno** [comment]
@richdouglasevans Thanks for the tip! I actually went a step further and used Array's built-in indexing (since I was just replicating anyway). RE: Hound, I'll leave it in for now, just trying it out, but I might tweak the settings.

**Comment by richdouglasevans** [comment]
> I actually went a step further and used Array's built-in indexing since I was just replicating anyway.

Ha! That makes even better sense. Johno, clearly you are a man of sagacious quality :+1:

**Pull request merged**